monotone.1
.TH "monotone" 1
.SH NAME
monotone \- distributed version control system
.SH SYNOPSIS
\fBmonotone\fP \fI[options] <command> [parameters]\fP
.P
Options, which affect global behavior or set default values, come
first in the argument list. A single command must follow, indicating
the operation to perform, followed by parameters which vary depending
on the command.
.SS Note
This manpage is a summary of some of the features and commands of
\fBmonotone\fP, but it is not the most detailed source of information
available. For a complete discussion of the concepts and a tutorial on
its use, please refer to the texinfo manual (via the \fBinfo
monotone\fP command, or online).
.SS Commands
.TP 
\fBcomment\fP \fI<id>\fP
Write a comment cert for a revision.
.TP
\fBapprove\fP \fI<id>\fP
Make a "branch" cert approving of a revision's membership in a branch.
.TP
\fBdisapprove\fP \fI<id1>\fP
Disapprove of a revision, committing the inverse changes as as a
descendant of the disapproved revision.
.TP
\fBtag\fP \fI<id> <tagname>\fP
Put a symbolic tag cert on a revision.
.TP
\fBtestresult\fP \fI<id> (0|1|true|false|yes|no|pass|fail)\fP
Indicate a passing or failing test result on a revision.
.TP
\fBdiff \fI[--revision=<id1> [--revision=<id2>] ] [<pathname>...]\fP
Show diffs between working copy and database.
.TP
\fBstatus \fI[<pathname>...]\fP
Show status of working copy.
.TP
\fBlog\fP \fI[id] \fP
Show historical log of revisions, starting from working copy
base revision, or \fI[id]\fP if given.
.TP
\fBcert\fP \fI<id> <certname> [certval]\fP
Create a custom cert for a revision. Reads cert value
from stdin if no value given on command line.
.TP
\fBgenkey\fP \fI<keyid>\fP
Generate an RSA key-pair and store it in the database.
.TP
\chkeypass\fP \fI<keyid>\fP
Change passphrase of the private half of a key.
.TP
\fBlist certs\fP \fI<id>\fP
List certs associated with revision.
.TP
\fBlist keys\fP \fI[partial-id]\fP
List keys matching glob, or list all keys if no glob given.
.TP
\fBlist branches\fP
List all branches.
.TP
\fBlist tags\fP
List all tags.
.TP
\fBlist vars \fI[<domain>]\fP
List all vars (possibly limited by domain).
.TP
\fBlist unknown \fI[<pathname...]\fP
List files in working directory, but not in revision's manifest or
work list.
.TP
\fBlist ignored \fI[<pathname...]\fP
List files intentionally ignored due to the ignore_file hook.
.TP
\fBlist missing \fI[<pathname...]\fP
List files in revision's manifest, but not in working directory.
.TP
\fBfdata\fP \fI<id>\fP
Write file data packet to stdout.
.TP
\fBfdelta\fP \fI<oldid> <newid>\fP
Write file delta packet to stdout.
.TP
\fBmdata\fP \fI<id>\fP
Write manifest data packet to stdout.
.TP
\fBmdelta\fP \fI<oldid> <newid>\fP
Write manifest delta packet to stdout.
.TP
\fBcerts\fP \fI<id>\fP
Write revision cert packets to stdout.
.TP
\fBrdata\fP \fI<id>\fP
Write revision data packet to stdout.
.TP
\fBprivkey\fP \fI<id>\fP
Write private key packet to stdout.
.TP
\fBpubkey\fP \fI<id>\fP
Write public key packet to stdout.
.TP
\fBread\fP \fI[<file1> [<file2> [...]]]\fP
Read packets from files or stdin. 
.TP
\fBcvs_import\fP \fI<cvsroot>/<module>\fP
Import all versions in CVS module. Reconstructs revisions and converts
metadata to certificates. A private signing key must already exist in
the database.
.TP
\fBrcs_import\fP \fI<rcsfile> ...\fP
Import all file versions in RCS files. Does not reconstruct revisions
across the entire tree.  You do not want this command, it is for
debugging; use cvs_import.
.TP
\fBcheckout\fP \fI[--revision=revision-id]\fP \fI[<directory>]\fP
Check out revision from database.
.TP
\fBco\fP \fI[--revision=revision-id]\fP \fI[<directory>]\fP
Check out revision from database; an alias for \fBcheckout\fP.
.TP
\fBcat\fP \fI(file|manifest) <id>\fP
Write file or manifest from database to stdout.
.TP
\fBheads\fP
Show unmerged heads of branch, or report when branch is merged.
.TP
\fBmerge\fP
Merge unmerged heads of branch.
.TP
\fBadd\fP \fI<pathname> [...]\fP
Add files to working copy. adding a file does not copy it into the database,
merely adds it to the work list. You must \fBcommit\fP your changes in order
to copy added files to the database.
.TP
\fBdrop\fP \fI<pathname> [...]\fP
Drop files from working copy. Files are not deleted from working copy, 
merely noted as removals in the work list.
.TP
\fBrename\fP \fI<src> \fI<dst>\fP
Rename files from \fI<src> \fP to \fI<dst> \fP in working copy.
.TP
\fBcommit\fP \fI[(--message=<log message>|--message-file=<file>)] [<pathname>...]\fP
Commit working copy to database.  Each commit has a changelog message
associated with it.  If --message is provided on the command line, it
is used; if --message-file is provided, the content of the
named file will be used as a commit message. If the filename is '-'
the commit message will be read from standard input.  Otherwise a log
message editor will be invoked.  If the file MT/log exists
and is non-empty, its content is used to prefill the editor.  You
cannot specify both --message and --message-file at the same time, and
if MT/log exists and is non-empty, you can cannot specify either of them
at all.
.TP
\fBupdate\fP  \fI[revision-id]\fP
Update working copy.
.TP
\fBrefresh_inodeprints\fP 
Turn on inodeprints mode, and force a cache refresh.
.TP
\fBpush\fP \fI[<host> [<glob>]]\fP 
Push contents of \fI<glob>\fP to database on \fI<host>\fP.
.TP
\fBpull\fP \fI[<host> [<glob>]]\fP 
Push contents of \fI<glob>\fP from database on \fI<host>\fP.
.TP
\fBsync\fP \fI<host> <glob>\fP 
Sync contents of \fI<glob>\fP with database on \fI<host>\fP.
.TP
\fBserve\fP \fI[--pid-file=<path>] <host>[:<port> <glob>\fP 
Serve contents of \fI<glob>\fP at network address \fI<host>\fP, on the
port \fI<port>\fP.  If \fI<port>\fP isn't given, 5253 is used. If a
--pid-file option is provided on the command line, monotone will store
the process id of the server in the specified file.
.TP
\fBset\fP \fI<domain> <name> <value>\fP 
Set the db var \fI<name>\fP in domain \fI<domain>\fP to value
\fI<value>\fP. 
.TP
\fBunset\fP \fI<domain> <name>\fP 
Delete any setting for db var \fI<name>\fP in domain \fI<domain>\fP.
.TP
\fBautomate\fP \fI(interface_version|heads|ancestors|attributes|parents|descendents|children|graph|erase_ancestors|toposort|ancestry_difference|leaves|inventory|stdio|certs|select)\fP
Scripting interface.
.TP
\fBdb\fP \fI(init|info|version|dump|load|migrate|rebuild|execute|check)\fP
Manipulate database state.
.SH DESCRIPTION
Monotone is a version control system, which allows you to keep old
versions of files, as well as special \fImanifest files\fP which
describe the location of files in a tree. Unlike other systems,
versions in monotone are \fIidentified\fP by cryptographic hash, and
operations are authenticated by individual users' evaluating
cryptographic signatures on meta-data, rather than any central
authority.

Monotone keeps a collection of versions in a single-file relational
database. It is essentially serverless, using network servers only as
untrusted communication facilities. A monotone database is a regular
file, which contains all the information needed to extract previous
versions of files, verify signatures, merge and modify versions, and
communicate with network servers.
.SH OPTIONS
.TP
\fB--help\fP
Print help message.
.TP
\fB--debug\fP
Turn on debugging log on standard error stream. This is very
verbose. Default is to be silent, unless an error occurs, in which
case failure log is dumped.
.TP
\fB--quiet\fP
Turn off normal progress messages.
.TP
\fB--dump=\fP\fI<file>\fP
Dump debugging log to \fIfile\fP on failure.
.TP
\fB--nostd\fP
Do not evaluate "standard" lua hooks compiled into \fBmonotone\fP.
.TP
\fB--norc\fP
Do not load lua hooks from user's \fB~/.monotonerc\fP file.
.TP
\fB--rcfile=\fP\fI<file>\fP
Load extra lua hooks from \fIfile\fP (may be given multiple times).
.TP
\fB--db=\fP\fI<file>\fP
Use database in \fIfile\fP.
.TP
\fB--key=\fP\fI<keyid>\fP
Use \fIkeyid\fP for operations which produce RSA signatures. Default
is inferred from presence of unique private key in database. Can also
be customized on a per-branch basis with hook function 
\fBget_branch_key(branchname)\fP.
.TP
\fB-k\fP \fI<keyid>\fP
An alias for \fB--key=\fP\fI<keyid>\fP
.TP
\fB--branch=\fP\fI<branchname>\fP
Use \fIbranchname\fP for operations on a branch. Default is inferred
in operations on existing branches (commit, update, etc).
.TP
\fB-b\fP \fI<branchname>\fP
An alias for \fB--branch=\fP\fI<branchname>\fP
.TP
\fB--ticket=\fP\fIdot|count\fP
Use the given method to print tickers.  The \fBcount\fP method prints
the count for each ticker on one line, incrementing the numbers in
place, while the \fBdot\fP method prints a continuous string of
characters (like some programs provide a progress line of dots).
The default is \fBcount\fP.
.TP
\fB--revision=\fP\fI<id>\fP
Used to specify one or more revision ids to various commands.
.TP
\fB-r\fP \fI<id>\fP
An alias for \fB--revision=\fP\fI<id>\fP
.TP
\fB--message=\fI<log message>\fP
Use the given message as the changelog when committing a new revision
rather than invoking the log message editor. Currently this option only
applies to the commit command but it may also apply to the comment
command in the future.
This option is alternative to --message-file
.TP
\fB-m \fI<log message>\fP
An alias for \fB--message=\fI<log message>\fP
.TP
\fB--message-file=\fI<message file>\fP
Use the content of the given file as the changelog when committing a
new revision rather than invoking the log message editor. If the passed filename is '-' the changelog message will be read from standard input.
Currently this option only applies to the commit command but it may also apply
to the comment command in the future.
This option is alternative to --message.
.TP
\fB--author=\fI<author email>\fP
Use the given author as the value of the "author" cert when committing
a new revision, rather than the default author.  Useful when
committing a patch on behalf of someone else, or when importing
history from another version control system.
.TP
\fB--date=\fI<date and time>\fP
Use the given given date and time as value of the "date" cert when
committing a new revision, rather than the current time.  Useful when
importing history from another version control system.
.TP
\fB--root=\fI<root dir>\fP
Stop the search for a working copy (containing the @file{MT} directory)
at the specified root directory rather than at the physical root of the
filesystem.
.TP
\fB--xargs=\fI<file>\fP
Inject the contents of the file in place among the command line
arguments.  This may be useful in case the command line would
otherwise become too long for your system.  This option can be used
more than once if needed.
.TP
\fB-@ \fI<file>\fP
An alias for \fB--xargs=\fI<file>\fP
.TP
\fB--format \fI<format string>\fP
Applies the specified format string at automate or annotate output.
The format string is made by a common part, applied to each revision,
and one or more optional changeset subformats, applied to changeset
data.
Changeset subformats are enclosed beetween braces and start with an
uppercase formatting specifier, e.g. an expression using the 'added
files' subformat is %A{...}
Everything \fIinside\fP the braces is repeated for every matching
changeset item.  For example, %A{add: %f\\n} prints every added file
prefixed by 'add: ' and followed by a newline.
 
Base formatting specifiers:

%a  : value of author certificate
.br
%sa : shortened value of author certificate
.br
%b  : value of branch certificate
.br
%sb : shortened value of branch certificate
.br
%d  : value of date certificate
.br
%sd : shortened value of date certificate
.br
%e  : value of comment certificate
.br
%i  : revision id
.br
%si : shortened revision id
.br
%l  : value of changelog certificate
.br
%r  : value of testresult certificate
.br
%t  : value of tag certificate
.br
%m  : manifest id
.br
%sm : shortened manifest id
.br
%+  : terminates formatting for the current revision. The next revision
will be formatted by the specifiers following %+
.br
%%  : the % character. It could also be obtained with '\\%' (see below)


The short forms have the following meanings:  For author, everything 
before the first @ symbol.  For branch, everything after the last '.' 
character.  For date, everything before the first 'T' character.  
For ids, the first 8 characters.  

Changeset formatting specifiers:

%P : ancestors
.br
%A : added files
.br
%D : deleted files
.br
%E : deleted dirs
.br
%R : renamed files
.br
%C : renamed dirs
.br
%M : modified files

%P,%A,%D,%E,%M supports the private specifier %f, substituted with the
file/directory/ancestor name/id. 
%R and %M supports the specifiers %f, current file/directory name and
%o, old file/directory name. 
 
Plus, all format string support the  \\n, \\r, \\a, \\t, \\b, \\f, \\v
modifiers with the usual c-like meaning.
\\\\, \\%, \\} are used to obtain the \\, % and } chars respectively.

The default format string is '%i\\n'. 

The specifier %+ is used to apply alternating format strings to
different revisions. When found, a %+ specifiers terminates the
formatting of current revision.
The next revision to be formatted will 'see' only the format string
following the %s.
To have a two column author output you will use a '%a %+ %a\\n' format
string.

.TP
\fB--xml\fP
Generate an xml document based on automate output and containing full
revision and changeset data.


.SH ENVIRONMENT
.TP
\fBEDITOR\fP
Used to edit comments, log messages, etc.
.TP
\fBVISUAL\fP
Used in preference to \fBEDITOR\fP, if set.
.SH FILES
.TP
\fB$HOME/.monotonerc\fP
A lua script, used as a customization file.
.SH NOTES
.IP \(bu
Command line options override environment variables and
settings in lua scripts (such as \fB.monotonerc\fP)
.SH "SEE ALSO"
\fBinfo monotone\fP
.SH BUGS
see http://savannah.nongnu.org/bugs/?group=monotone
.SH AUTHOR
graydon hoare <graydon@pobox.com>


NEWS
mtsh 0.2	2005-10-09
	The file list now shows a proper tree, instead of a flat list.
	
	mtsh now has primitive "undrop" support
	
	The "File diff" and "File contents" tabs now get reset when a
	file that they can't be generated for is selected. Previously, they'd
	keep showing the contents of the previous file.

mtsh 0.1x
	Fix annoying crashes, and make the file list a drag&drop source.

mtsh 0.1
	Initial release.


monotone.cc
// Copyright 2005   Timothy Brownawell   <tbrownaw@gmail.com>
// Licensed to the public under the GNU GPL v2, see COPYING

#include "monotone.hh"

#include <sys/types.h>
#include <sys/time.h>
#include <sys/wait.h>
#include <sys/select.h>
#include <unistd.h>
#include <cstring>
#include <cerrno>
#include <csignal>

#include <map>

#include <boost/lexical_cast.hpp>

// For future reference: CreateProcess and WaitForMultipleObjects on win32
// to replace fork/exec and select?
// Put in a 'class Platform {};'?

inline int max(int a, int b) {return (a>b)?a:b;}

monotone::monotone(): pid(-1), lwcb(0)
{
}

monotone::~monotone()
{
  stop();
}

void
monotone::set_longwait_callback(longwait_callback lc)
{
  lwcb = lc;
}

bool
monotone::execute(std::vector<std::string> args)
{
  int to_mtn[2];
  int from_mtn[2];
  int err_mtn[2];
  if (pipe(to_mtn) < 0 || pipe(from_mtn) < 0 || pipe(err_mtn) < 0)
    return false;
  pid = fork();
  if (pid < 0)
    {
      close(to_mtn[0]);
      close(to_mtn[1]);
      close(from_mtn[0]);
      close(from_mtn[1]);
      close(err_mtn[0]);
      close(err_mtn[1]);
      return false;
    }
  else if (pid > 0)
    {
      close(to_mtn[0]);
      close(from_mtn[1]);
      close(err_mtn[1]);
      to = to_mtn[1];
      from = from_mtn[0];
      errfrom = err_mtn[0];
      return true;
    }
  else
    {
      if (!db.empty())
        args.push_back("--db=" + db);
      chdir(dir.c_str());
      if (close(to_mtn[1]) < 0 || close(from_mtn[0]) < 0)
        return false;
      if (close(0) < 0 || close(1) < 0 || close(2) < 0)
        return false;
      if (dup2(to_mtn[0], 0) < 0
          || dup2(from_mtn[1], 1) < 0
          || dup2(err_mtn[1], 2) < 0)
        return false;
      char **arg = new char *[args.size() + 2];
      arg[0] = "monotone";
      for (unsigned int i = 0; i < args.size(); ++i)
        {
          arg[i+1] = new char[args[i].size()+1];
          memcpy(arg[i+1], args[i].c_str(), args[i].size()+1);
        }
      arg[args.size()+1] = 0;
      int ret = execvp(arg[0], arg);
      exit(1);
    }
}
bool
monotone::start()
{
  stop();
  std::vector<std::string> args;
  args.push_back("automate");
  args.push_back("stdio");
  return execute(args);
}
bool
monotone::stop()
{
  if (pid == -1)
    return false;
  kill(pid, SIGKILL);
  int r;
  do {r = waitpid(pid, 0, 0);} while (r==-1 && errno == EINTR);
  pid = -1;
  return true;
}
bool
monotone::stopped()
{
  if (lwcb)
  	lwcb();
  if (pid == -1)
    return true;
  int r = waitpid(pid, 0, WNOHANG);
  if (r == 0)
    return false;
  pid = -1;
  return true;
}

bool
monotone::read_header(int & cmdnum, int & err, bool & more, int & size)
{
  char head[32];
  int got = 0;
  int minsize = 8;
  int c1(0), c2(0), c3(0), c4(0);
  while (!stopped() && !c4)
    {
      int r = read(from, head+got, minsize-got);
      got += r;
      if (!c1)
        for (int i = 0; i < got; ++i)
          if (head[i] == ':')
            c1 = i, i=got;
      if (!c2)
        for (int i = c1+1; i < got; ++i)
          if (head[i] == ':')
            c2 = i, i=got;
      if (!c3)
        for (int i = c2+1; i < got; ++i)
          if (head[i] == ':')
            c3 = i, i=got;
      if (!c4)
        for (int i = c3+1; i < got; ++i)
          if (head[i] == ':')
            c4 = i, i=got;
      if (c4)
        minsize = c4;
      else if (c3)
        minsize = max(c3+2, got+1);
      else if (c2)
        minsize = max(c2+4, got+3);
      else if (c1)
        minsize = max(c1+6, got+5);
      else
        minsize = max(8, got+7);
    }
  if (stopped())
    return false;
  cmdnum = boost::lexical_cast<int>(std::string(head, c1));
  err = boost::lexical_cast<int>(std::string(head+c1+1, c2-c1-1));
  more = (head[c2+1] == 'm');
  size = boost::lexical_cast<int>(std::string(head+c3+1, c4-c3-1));
  return true;
}

bool
monotone::read_packet(std::string & out)
{
  int cn, er, size;
  bool m = false;
  if (!read_header(cn, er, m, size))
    return false;
  char output[size];
  int got = 0;
  do
    {
      int r = read(from, output, size - got);
      got += r;
      out += std::string(output, r);
    } while (!stopped() && got != size);
  if (stopped())
    return false;
  return m;
}
std::string
monotone::command(std::string const & cmd,
                    std::vector<std::string> const & args)
{
  if (pid == -1)
    start();
  std::string res;
  std::ostringstream s;
  s << cmd.size() <<":"<<cmd;
  for (std::vector<std::string>::const_iterator i = args.begin();
        i != args.end(); ++i)
    s << i->size() << ":" << *i;
  std::string c = "l" + s.str() + "e";
  write(to, c.c_str(), c.size());
  while(!stopped() && read_packet(res))
    ;
  return res;
}

void
monotone::runcmd(std::string const & cmd,
                 std::vector<std::string> const & args,
                 std::string & out, std::string & err)
{
  out.clear();
  err.clear();
  stop();// stop stdio
  std::vector<std::string> argg = args;
  argg.insert(argg.begin(), cmd);
  execute(argg);

  int size = 2048;
  char *output = new char[size];
  char *errout = new char[size];
  fd_set rd, ex;
  do
    {
      timeval timeout;
      timeout.tv_sec = 0;
      timeout.tv_usec = 100000;
      FD_ZERO(&rd);
      FD_ZERO(&ex);
      FD_SET(from, &rd);
      FD_SET(errfrom, &rd);
      FD_SET(from, &ex);
      FD_SET(errfrom, &ex);
      int s = ::select(max(from, errfrom)+1, &rd, 0, &ex, &timeout);
      if (FD_ISSET(from, &rd))
        {
          int r = read(from, output, size);
          out += std::string(output, r);
        }
      if (FD_ISSET(errfrom, &rd))
        {
          int r = read(errfrom, errout, size);
          err += std::string(errout, r);
        }
    } while (!stopped());
  int r = read(from, output, size);
  out += std::string(output, r);
  r = read(errfrom, errout, size);
  err += std::string(errout, r);
  delete[] output;
  delete[] errout;
}

void
monotone::inventory(std::vector<inventory_item> & out)
{
  out.clear();
  std::vector<std::string> args;
  std::string res = command("inventory", args);
  std::map<int, int> renames;
  int begin = 0;
  int end = res.find('\n', begin);
  while (begin != res.size())
    {
      int sp1 = begin + 4;
      int sp2 = res.find(' ', sp1 + 1);
      int sp3 = res.find(' ', sp2 + 1);
      int fromid = boost::lexical_cast<int>(res.substr(sp1, sp2-sp1));
      int toid = boost::lexical_cast<int>(res.substr(sp2+1, sp3-sp2-1));
      std::string path = res.substr(sp3+1, end - sp3-1);
      inventory_item *pre, *post;
      if (!fromid || !toid)
        out.push_back(inventory_item());
      if (!fromid)
        pre = &out.back();
      if (!toid)
        post = &out.back();
      if (fromid)
        {
          std::map<int, int>::iterator i = renames.find(fromid);
          if (i != renames.end())
            pre = &out[i->second];
          else
            {
              renames.insert(std::make_pair(fromid, out.size()));
              out.push_back(inventory_item());
              pre = &out.back();
            }
        }
      if (toid)
        {
          std::map<int, int>::iterator i = renames.find(toid);
          if (i != renames.end())
            post = &out[i->second];
          else
            {
              renames.insert(std::make_pair(toid, out.size()));
              out.push_back(inventory_item());
              post = &out.back();
            }
        }
      switch (res[begin])//prestate
        {
        case 'D':
        case 'R':
          pre->prename = path;
        default:
          ;
        }
      switch (res[begin+1])//poststate
        {
        case 'R':
        case 'A':
          post->postname = path;
          break;
        default:
          if (pre->prename.empty())
            {
              post->postname = path;
              pre->prename = path;
            }
        }
      switch (res[begin+2])//filestate
        {
        case 'M':
          post->state = inventory_item::missing;
          break;
        case 'P':
          post->state = inventory_item::patched;
          break;
        case 'I':
          if (!fromid && !toid)
            post->state = inventory_item::ignored;
          break;
        case 'U':
          if (!fromid && !toid)
            post->state = inventory_item::unknown;
          break;
        default:
          ;
        }
      begin = end + 1;
      end = res.find('\n', begin);
    }
}
std::vector<cert>
monotone::certs(std::string const & rev)
{
  std::vector<std::string> args;
  std::vector<cert> out;
  args.push_back(rev);
  std::string res = command("certs", args);
  cert c;
  for (int begin = 0, end = res.find('\n'); begin != res.size();
        begin = end + 1, end = res.find('\n', begin))
    {
      std::string line = res.substr(begin, end-begin);
      int lpos = line.find('"');// other is at end of line
      if (lpos == std::string::npos)
        {
          out.push_back(c);
          continue;
        }
      while (line[line.size()-1] != '"' || line[line.size()-2] =='\\')
        {
          begin = end + 1;
          end = res.find('\n', begin);
          line += "\n" + res.substr(begin, end-begin);
        }
      std::string contents;
      for (int i = lpos+1; i < line.size() && line[i] != '"'; ++i)
        {
          if (line[i] == '\\')
            ++i;
          contents += line[i];
        }
      if (line.find("key") < lpos)
        c.key = contents;
      else if (line.find("signature") < lpos)
        c.sig = (contents == "ok");
      else if (line.find("name") < lpos)
        c.name = contents;
      else if (line.find("value") < lpos)
        c.value = contents;
      else if (line.find("trust") < lpos)
        c.trusted = (contents == "trusted");
    }
  if (res.size())
    out.push_back(c);
  return out;
}
std::vector<std::string>
monotone::select(std::string const & sel)
{
  std::vector<std::string> args, out;
  args.push_back(sel);
  std::string res = command("select", args);
  int begin = 0;
  int end = res.find('\n', begin);
  while (begin != res.size())
    {
      out.push_back(res.substr(begin, end-begin));
      begin = end + 1;
      end = res.find('\n', begin);
    }
  return out;
}
void
monotone::make_cert(std::string const & rev,
                std::string const & name,
                std::string const & value)
{
  std::vector<std::string> args;
  args.push_back(rev);
  args.push_back(name);
  args.push_back(value);
  std::string ign1, ign2;
  runcmd("cert", args, ign1, ign2);
}
std::string
monotone::commit(std::vector<std::string> args)
{
  std::string ign1, ign2;
  runcmd("commit", args, ign1, ign2);
  args.clear();
  std::string res = command("get_revision", args);
  for (int begin = 0, end = res.find('\n'); begin != res.size();
        begin = end + 1, end = res.find('\n', begin))
    {
      std::string line = res.substr(begin, end-begin);
      int lpos = line.find_first_of("[\"");
      int rpos = line.find_first_of("]\"", lpos + 1);
      std::string contents = line.substr(lpos + 1, rpos - lpos - 1);
      if (line.find("old_revision") < lpos)
        return contents;
    }
  return std::string();
}
std::string
monotone::diff(std::string const & filename)
{
  std::string out, ign;
  std::vector<std::string> args;
  args.push_back(filename);
  runcmd("diff", args, out, ign);
  return out;
}
std::string
monotone::diff(std::string const & filename,
                  std::string const & rev1,
                  std::string const & rev2)
{
  std::vector<std::string> args;
  args.push_back(filename);
  args.push_back("--revision=" + rev1);
  args.push_back("--revision=" + rev2);
  std::string out, ign;
  runcmd("diff", args, out, ign);
  return out;
}
std::string
monotone::cat(std::string const & filename, std::string const & rev)
{
  std::vector<std::string> args;
  args.push_back(filename);
  args.push_back("--revision=" + rev);
  std::string out, ign;
  runcmd("cat", args, out, ign);
  return out;
}
std::string
monotone::get_revision(std::string const & rev)
{
  std::vector<std::string> args;
  args.push_back(rev);
  return command("get_revision", args);
}
std::string
monotone::get_manifest(std::string const & rev)
{
  std::vector<std::string> args;
  args.push_back(rev);
  return command("get_manifest", args);
}
void
monotone::add(std::string const & file)
{
  std::string ign1, ign2;
  std::vector<std::string> args;
  args.push_back(file);
  runcmd("add", args, ign1, ign2);
}
void
monotone::drop(std::string const & file)
{
  std::string ign1, ign2;
  std::vector<std::string> args;
  args.push_back(file);
  runcmd("drop", args, ign1, ign2);
}
void
monotone::revert(std::string const & file)
{
  std::string ign1, ign2;
  std::vector<std::string> args;
  args.push_back(file);
  runcmd("revert", args, ign1, ign2);
}
void
monotone::rename(std::string const & oldname, std::string const & newname)
{
  std::string ign1, ign2;
  std::vector<std::string> args;
  args.push_back("--execute");
  args.push_back(oldname);
  args.push_back(newname);
  runcmd("rename", args, ign1, ign2);
}

bool
monotone::update(std::vector<std::string> & opts)
{
  opts.clear();
  std::string ign, err;
  std::vector<std::string> args;
  runcmd("update", args, ign, err);
  int p = err.find("multiple update candidates");
  if (p == std::string::npos)
    return true;

  p = err.find("monotone:   ", p + 1);
  while (p != std::string::npos)
    {
      p = err.find(":", p);
      p = err.find_first_not_of(" ", p + 1);
      opts.push_back(err.substr(p, 40));
      p = err.find("monotone:   ", p + 1);
    }
  return false;
}

void
monotone::update(std::string const & rev)
{
  std::string ign, err;
  std::vector<std::string> args;
  args.push_back("--revision="+rev);
  runcmd("update", args, ign, err);
}
